README
NAME
    RT::Extension::ExternalStorage - Store attachments outside the database

SYNOPSIS
        Set( @Plugins, 'RT::Extension::ExternalStorage' );

        Set(%ExternalStorage,
            Type => 'Disk',
            Path => '/opt/rt4/var/attachments',
        );

DESCRIPTION
    By default, RT stores attachments in the database. This extension moves
    all attachments that RT does not need efficient access to (which include
    textual content and images) to outside of the database. This may either
    be on local disk, or to a cloud storage solution. This decreases the
    size of RT's database, in turn decreasing the burden of backing up RT's
    database, at the cost of adding additional locations which must be
    configured or backed up.

    The files are initially stored in the database when RT receives them;
    this guarantees that the user does not need to wait for the file to be
    transferred to disk or to the cloud, and makes it durable to transient
    failures of cloud connectivity. The provided bin/extract-attachments
    script, to be run regularly via cron, takes care of moving attachments
    out of the database at a later time.

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        If you are using RT 4.2 or greater, add this line:

            Plugin('RT::Extension::ExternalStorage');

        For RT 4.0, add this line:

            Set(@Plugins, qw(RT::Extension::ExternalStorage));

        or add RT::Extension::ExternalStorage to your existing @Plugins
        line.

        You will also need to configure the %ExternalStorage option,
        depending on how and where you want your data stored; see
        "CONFIGURATION".

    Restart your webserver
        Restarting the webserver before the next step (extracting existing
        attachments) is important to ensure that files remain available as
        they are extracted.

    Extract existing attachments
        Run bin/extract-attachments; this may take some time, depending on
        the existing size of the database. This task may be safely cancelled
        and re-run to resume.

    Schedule attachments extraction
        Schedule bin/extract-attachments to run at regular intervals via
        cron. For instance, the following /etc/cron.d/rt entry will run it
        daily, which may be good to concentrate network or disk usage to
        times when RT is less in use:

            0 0 * * * root /opt/rt4/local/plugins/RT-Extension-ExternalStorage/bin/extract-attachments

CONFIGURATION
    This module comes with a number of possible backends; see the
    documentation in each for necessary configuration details:

    RT::Extension::ExternalStorage::Disk
    RT::Extension::ExternalStorage::Dropbox
    RT::Extension::ExternalStorage::AmazonS3

AUTHOR
    Best Practical Solutions, LLC <modules@bestpractical.com>

BUGS
    All bugs should be reported via email to

        L<bug-RT-Extension-ExternalStorage@rt.cpan.org|mailto:bug-RT-Extension-ExternalStorage@rt.cpan.org>

    or via the web at

        L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-ExternalStorage>.

COPYRIGHT
    This extension is Copyright (C) 2009-2015 Best Practical Solutions, LLC.

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
0.62 2016-10-07
 - Store backend-specific keys in the attachment

0.61 2015-02-25
 - Documentation additions; note that content is de-duplicated, and that
   shredder will not reclaim space

0.60 2015-01-20
 - Never store 0-length attachments of OCFVs externally

0.55 2015-01-05
 - Update copyright year
 - Add explicit "return 0"s for non-external attachment storage
 - Store large (>10M) CFs of type Image externally, as well

0.51 2014-12-18
 - Export old file upload CFs, as well as current values
 - Add no_index for core RT packages

0.50 2014-12-10
 - Initial release



lib/RT/Extension/ExternalStorage/Box.pm
# vim: ai ts=4 sts=4 et sw=4 ft=perl
# BEGIN BPS TAGGED BLOCK {{{
#
# COPYRIGHT:
#
# This software is Copyright (c) 1996-2016 Best Practical Solutions, LLC
#                                          <sales@bestpractical.com>
#
# (Except where explicitly superseded by other copyright notices)
#
#
# LICENSE:
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from www.gnu.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to Best Practical Solutions, LLC.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# Request Tracker, to Best Practical Solutions, LLC, you confirm that
# you are the copyright holder for those contributions and you grant
# Best Practical Solutions,  LLC a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# END BPS TAGGED BLOCK }}}

use 5.008003;
use warnings;
use strict;

package RT::Extension::ExternalStorage::Box;

use Role::Basic qw/with/;
with 'RT::Extension::ExternalStorage::Backend';

our ( $Box );

sub Init {
    my $self = shift;
    my %self = %{$self};

    if (not Storage::Box->require) {
        RT->Logger->error("Required module Storage::Box is not installed");
        return;
    } elsif (not $self{KeyId}) {
        RT->Logger->error("KeyId not provided for Box.com");
        return;
    } elsif (not $self{EnterpriseId}) {
        RT->Logger->error("EnterpriseId not provided for Box.com");
        return;
    } elsif (not $self{PrivateKey}) {
        RT->Logger->error("PrivateKey not provided for Box.com");
        return;
    } elsif (not $self{Password}) {
        RT->Logger->error("Password not provided for Box.com");
        return;
    } elsif (not $self{ClientId}) {
        RT->Logger->error("ClientId not provided for Box.com");
        return;
    } elsif (not $self{ClientSecret}) {
        RT->Logger->error("ClientSecret not provided for Box.com");
        return;
    }

    $Box = Storage::Box->new(
        key_id => $self{KeyId},
        enterprise_id => $self{EnterpriseId},
        private_key => $self{PrivateKey},
        password => $self{Password},
        client_id => $self{ClientId},
        client_secret => $self{ClientSecret}
    );

    # the first time we run, we create a 'rt' app user at Box.com
    # this user id get stored for all future file access
    # if this id is lost, it will create a new rt user but
    # that user won't have access to any of the files!!!
    my $BoxUserId = RT->System->FirstAttribute("BoxUserId");
    my $user_id = $BoxUserId ? $BoxUserId->Content || '' : '';
    if ($user_id eq '') {
        $user_id = $Box->create_user('rt');
	RT->Logger->info("Box.com rt user id $user_id");
        RT->System->SetAttribute(
            Name => "BoxUserId",
            Description => "User ID for the Box.com rt user",
            Content => $user_id
        );
    }
    $Box->user_id($user_id);

    return $self;
}

sub Get {
    my ($self,$key) = @_;
    RT->Logger->info("Downloading $key");
    my $contents = $Box->download_file($key);
    $contents;
}

sub Store {
    my ($self,$key,$content) = @_;

    RT->Logger->info("Box Storing $key");
    # we need to store the file locally for libcurl to be able to upload it
    # this is a limitation of WWW::Curl::Form module, as it lacks support
    # for the CURLFORM_BUFFER and CURLFORM_BUFFERPTR options, rather it only
    # supports CURLFORM_FILE and CURLFORM_FILENAME.  But atleast it will upload
    # large files.
    open( my $fh, ">:raw", $key ) or return (undef, "Cannot write file to disk: $!");
    print $fh $content or return (undef, "Cannot write file to disk: $!");
    close $fh or return (undef, "Cannot write file to disk: $!");

    RT->Logger->info("Created file $key");
    my $file = $Box->create_file($key);
    RT->Logger->info("Created file " . $file);

    unlink $key;        # delete the file so we don't litter
    return ($file); # we return the file_id for get to get it
}

=pod

=head1 NAME

RT::Extension::ExternalStorage::Box

=head1 SYNOPSIS

	Plugin('RT::Extension::ExternalStorage');

	Set(%ExternalStorage,
		Type => 'Box',
		KeyId => "box_com_key_id",
		EnterpriseId =>  'box_com_enterprise_id',
		PrivateKey => "/opt/rt4/etc/keys/private_key.pem",
		Password => "my_secret_password",
		ClientId =>  "box_com_client_id",
		ClientSecret => "box_com_client_secret");

=head1 DESCRIPTION

The C<RT::Extension::ExternalStorage::Box> package provides an interface
that allows RT to store large attachments in Box.com's file storage.  It
requires setting up a custom enterprise application at Box.com attached
to your enterprise account in order to enable creating an application
managed Box.com user for RT, and provide a JWT based access credentials
for RT.

=head1 INSTALLATION


=over

=item 1.

Install storage-box and rt-extension-externalstorage:

    cpanm Storage::Box RT::Extension::ExternalStorage

=item 2.

Signup for an account at box.

=item 3.

Login to developer.box.com or L<https://app.box.com/developers/services>

=item 4.

Click Get Started if this is your first application

=item 5.

Create a unique name for your app, for example rt-myorganization

=item 6.

Under OAuth parameters, copy C<client_id> and C<client_secret> somewhere safe for later use

=item 7.

Add a redirect uri, it must be https, but need not exist.  We won't be using it anyways.

=item 8.

Under C<Authentication Type> select C<Server Authentication (OAuth2.0 with JWT)>

=item 9.

Under Scopes, Enterprise select C<Manage app users>

=item  10.

Before you can enable Public Key Managment, under C<Settings /  Security > select C<Login verification: Require 2-step verification for unrecognized logins>

=item 11.

Go back to your app and under Public Key Management, and select C<Add Public Key>

=item 12.

Using openssl, generate a private key with password in pem format by:

	openssl genrsa -aes256 -out private_key.pem 2048

=item 13.

Using openssl, create the corresponding public key file:

	openssl rsa -pubout -in private_key.pem -out public_key.pem

=item 14.

Save your password where you put the client_id and client_secret!

=item 15.

Copy and paste your public key into the C<Public Key> box and click C<Verify> and then C<Save>, you may have to enter your F2A credentials again after this.

=item 16.

Copy the Key ID next to Public Key 1 to the same safe place you are keeping your other secrets.

=item 17.

At the bottom of the page click C<Save Applications>

=item 18.

create a directory for you private and public key in the rt4 install directory such as:

	mkdir -p /opt/rt4/etc/keys
	mv *.pem /opt/rt4/etc/keys/

=item 19.

Under C<Settings / Business Settings> find the field C<Enterprise ID> and copy that to your list of secrets.

=item 20.

Edit your RT_SiteConfig.pm file to enable the C<Box> backend using the values you've saved in a safe place:

	Plugin('RT::Extension::ExternalStorage');

	Set(%ExternalStorage,
		Type => 'Box',
		KeyId => "KEY ID FROM STEP 16",
		EnterpriseId =>  'ENTERPRISE ID FROM STEP 19',
		PrivateKey => "/opt/rt4/etc/keys/private_key.pem",
		Password => "PASSWORD FROM STEP 14",
		ClientId =>  "CLIENT ID FROM STEP 6",
		ClientSecret => "CLIENT SECRET FROM STEP 6");

=item 21.

Trial run

Assuming your private key is installed and readable by your webserver process you should now have the integration working.  Running:

	/opt/rt4/local/plugins/RT-Extension-ExternalStorage/sbin/extract-attachments

by hand should copy the large files out of the current database and migrate them to Box.com.

=item 22.

Add that script to a cron job such as:

	 0 0 * * * root /opt/rt4/local/plugins/RT-Extension-ExternalStorage/sbin/extract-attachments

=back

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

All bugs should be reported via email to

L<bug-RT-Extension-ExternalStorage@rt.cpan.org|mailto:bug-RT-Extension-ExternalStorage@rt.cpan.org>

or via the web at

L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-ExternalStorage>.

=head1 COPYRIGHT

This extension is Copyright (C) 2009-2015 Best Practical Solutions, LLC.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

1;



lib/RT/Extension/ExternalStorage/Disk.pm
# BEGIN BPS TAGGED BLOCK {{{
#
# COPYRIGHT:
#
# This software is Copyright (c) 1996-2015 Best Practical Solutions, LLC
#                                          <sales@bestpractical.com>
#
# (Except where explicitly superseded by other copyright notices)
#
#
# LICENSE:
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from www.gnu.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to Best Practical Solutions, LLC.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# Request Tracker, to Best Practical Solutions, LLC, you confirm that
# you are the copyright holder for those contributions and you grant
# Best Practical Solutions,  LLC a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# END BPS TAGGED BLOCK }}}

use 5.008003;
use warnings;
use strict;

package RT::Extension::ExternalStorage::Disk;

use File::Path qw//;

use Role::Basic qw/with/;
with 'RT::Extension::ExternalStorage::Backend';

sub Init {
    my $self = shift;

    my %self = %{$self};
    if (not $self{Path}) {
        RT->Logger->error("No path provided for local storage");
        return;
    } elsif (not -e $self{Path}) {
        RT->Logger->error("Path provided for local storage ($self{Path}) does not exist");
        return;
    } elsif ($self{Write} and not -w $self{Path}) {
        RT->Logger->error("Path provided for local storage ($self{Path}) is not writable");
        return;
    }

    return $self;
}

sub Get {
    my $self = shift;
    my ($sha) = @_;

    $sha =~ m{^(...)(...)(.*)};
    my $path = $self->{Path} . "/$1/$2/$3";

    return (undef, "File does not exist") unless -e $path;

    open(my $fh, "<", $path) or return (undef, "Cannot read file on disk: $!");
    my $content = do {local $/; <$fh>};
    $content = "" unless defined $content;
    close $fh;

    return ($content);
}

sub Store {
    my $self = shift;
    my ($sha, $content) = @_;

    $sha =~ m{^(...)(...)(.*)};
    my $dir  = $self->{Path} . "/$1/$2";
    my $path = "$dir/$3";

    return ($sha) if -f $path;

    File::Path::make_path($dir, {error => \my $err});
    return (undef, "Making directory failed") if @{$err};

    open( my $fh, ">:raw", $path ) or return (undef, "Cannot write file on disk: $!");
    print $fh $content or return (undef, "Cannot write file to disk: $!");
    close $fh or return (undef, "Cannot write file to disk: $!");

    return ($sha);
}

=head1 NAME

RT::Extension::ExternalStorage::Disk - On-disk storage of attachments

=head1 SYNOPSIS

    Set(%ExternalStorage,
        Type => 'Disk',
        Path => '/opt/rt4/var/attachments',
    );

=head1 DESCRIPTION

This storage option places attachments on disk under the given C<Path>,
uncompressed.  The files are de-duplicated when they are saved; as such,
if the same file appears in multiple transactions, only one copy will be
stored on disk.

The C<Path> must be readable by the webserver, and writable by the
C<bin/extract-attachments> script.  Because the majority of the
attachments are in the filesystem, a simple database backup is thus
incomplete.  It is B<extremely important> that I<backups include the
on-disk attachments directory>.

Files also C<must not be modified or removed>; doing so may cause
internal inconsistency.

=cut

1;


lib/RT/Extension/ExternalStorage/Backend.pm
use 5.008003;
use warnings;
use strict;

package RT::Extension::ExternalStorage::Backend;

use Role::Basic;

requires 'Init';
requires 'Get';
requires 'Store';

sub new {
    my $class = shift;
    my %args = @_;

    $class = delete $args{Type};
    if (not $class) {
        RT->Logger->error("No storage engine type provided");
        return undef;
    } elsif ($class->require) {
    } else {
        my $long = "RT::Extension::ExternalStorage::$class";
        if ($long->require) {
            $class = $long;
        } else {
            RT->Logger->error("Can't load external storage engine $class: $@");
            return undef;
        }
    }

    unless ($class->DOES("RT::Extension::ExternalStorage::Backend")) {
        RT->Logger->error("External storage engine $class doesn't implement RT::Extension::ExternalStorage::Backend");
        return undef;
    }

    my $self = bless \%args, $class;
    $self->Init;
}

1;


lib/RT/Extension/ExternalStorage/Dropbox.pm
# BEGIN BPS TAGGED BLOCK {{{
#
# COPYRIGHT:
#
# This software is Copyright (c) 1996-2015 Best Practical Solutions, LLC
#                                          <sales@bestpractical.com>
#
# (Except where explicitly superseded by other copyright notices)
#
#
# LICENSE:
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from www.gnu.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to Best Practical Solutions, LLC.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# Request Tracker, to Best Practical Solutions, LLC, you confirm that
# you are the copyright holder for those contributions and you grant
# Best Practical Solutions,  LLC a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# END BPS TAGGED BLOCK }}}

use 5.008003;
use warnings;
use strict;

package RT::Extension::ExternalStorage::Dropbox;

use Role::Basic qw/with/;
with 'RT::Extension::ExternalStorage::Backend';

our $DROPBOX;
sub Init {
    my $self = shift;
    my %self = %{$self};

    if (not File::Dropbox->require) {
        RT->Logger->error("Required module File::Dropbox is not installed");
        return;
    } elsif (not $self{AccessToken}) {
        RT->Logger->error("AccessToken not provided for Dropbox.  Register a new application"
                      . " at https://www.dropbox.com/developers/apps and generate an access token.");
        return;
    }


    $DROPBOX = File::Dropbox->new(
        oauth2       => 1,
        access_token => $self{AccessToken},
        root         => 'sandbox',
        furlopts     => { timeout => 60 },
    );

    return $self;
}

sub Get {
    my $self = shift;
    my ($sha) = @_;

    open( $DROPBOX, "<", $sha)
        or return (undef, "Failed to retrieve file from dropbox: $!");
    my $content = do {local $/; <$DROPBOX>};
    close $DROPBOX;

    return ($content);
}

sub Store {
    my $self = shift;
    my ($sha, $content) = @_;

    # No-op if the path exists already.  This forces a metadata read.
    return ($sha) if open( $DROPBOX, "<", $sha);

    open( $DROPBOX, ">", $sha )
        or return (undef, "Open for write on dropbox failed: $!");
    print $DROPBOX $content
        or return (undef, "Write to dropbox failed: $!");
    close $DROPBOX
        or return (undef, "Flush to dropbox failed: $!");

    return ($sha);
}

=head1 NAME

RT::Extension::ExternalStorage::Dropbox - Store files in the Dropbox cloud

=head1 SYNOPSIS

    Set(%ExternalStorage,
        Type => 'Dropbox',
        AccessToken => '...',
    );

=head1 DESCRIPTION

This storage option places attachments in the Dropbox shared file
service.  The files are de-duplicated when they are saved; as such, if
the same file appears in multiple transactions, only one copy will be
stored on in Dropbox.

Files in Dropbox C<must not be modified or removed>; doing so may cause
internal inconsistency.  It is also important to ensure that the Dropbox
account used has sufficient space for the attachments, and to monitor
its space usage.

=head1 SETUP

In order to use this stoage type, a new application must be registered
with Dropbox:

=over

=item 1.

Log into Dropbox as the user you wish to store files as.

=item 2.

Click C<Create app> on L<https://www.dropbox.com/developers/apps>

=item 3.

Choose B<Dropbox API app> as the type of app.

=item 4.

Choose the B<Files and datastores> as the type of data to store.

=item 5.

Choose B<Yes>, your application only needs access to files it creates.

=item 6.

Enter a descriptive name -- C<Request Tracker files> is fine.

=item 7.

Under C<Generated access token>, click the C<Generate> button.

=item 8.

Copy the provided value into your F<RT_SiteConfig.pm> file as the
C<AccessToken>:

    Set(%ExternalStorage,
        Type => 'Dropbox',
        AccessToken => '...',   # Replace the value here, between the quotes
    );

=back

=cut

1;


lib/RT/Extension/ExternalStorage.pm
# BEGIN BPS TAGGED BLOCK {{{
#
# COPYRIGHT:
#
# This software is Copyright (c) 1996-2015 Best Practical Solutions, LLC
#                                          <sales@bestpractical.com>
#
# (Except where explicitly superseded by other copyright notices)
#
#
# LICENSE:
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from www.gnu.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to Best Practical Solutions, LLC.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# Request Tracker, to Best Practical Solutions, LLC, you confirm that
# you are the copyright holder for those contributions and you grant
# Best Practical Solutions,  LLC a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# END BPS TAGGED BLOCK }}}

use 5.008003;
use warnings;
use strict;

package RT::Extension::ExternalStorage;

our $VERSION = '0.62';

use Digest::SHA qw//;

require RT::Extension::ExternalStorage::Backend;

=head1 NAME

RT::Extension::ExternalStorage - Store attachments outside the database

=head1 SYNOPSIS

    Set( @Plugins, 'RT::Extension::ExternalStorage' );

    Set(%ExternalStorage,
        Type => 'Disk',
        Path => '/opt/rt4/var/attachments',
    );

=head1 DESCRIPTION

By default, RT stores attachments in the database.  This extension moves
all attachments that RT does not need efficient access to (which include
textual content and images) to outside of the database.  This may either
be on local disk, or to a cloud storage solution.  This decreases the
size of RT's database, in turn decreasing the burden of backing up RT's
database, at the cost of adding additional locations which must be
configured or backed up.  Attachment storage paths are calculated based
on file contents; this provides de-duplication.

The files are initially stored in the database when RT receives them;
this guarantees that the user does not need to wait for the file to be
transferred to disk or to the cloud, and makes it durable to transient
failures of cloud connectivity.  The provided C<bin/extract-attachments>
script, to be run regularly via cron, takes care of moving attachments
out of the database at a later time.

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt4/etc/RT_SiteConfig.pm>

If you are using RT 4.2 or greater, add this line:

    Plugin('RT::Extension::ExternalStorage');

For RT 4.0, add this line:

    Set(@Plugins, qw(RT::Extension::ExternalStorage));

or add C<RT::Extension::ExternalStorage> to your existing C<@Plugins> line.

You will also need to configure the C<%ExternalStorage> option,
depending on how and where you want your data stored; see
L</CONFIGURATION>.

=item Restart your webserver

Restarting the webserver before the next step (extracting existing
attachments) is important to ensure that files remain available as they
are extracted.

=item Extract existing attachments

Run C<bin/extract-attachments>; this may take some time, depending on
the existing size of the database.  This task may be safely cancelled
and re-run to resume.

=item Schedule attachments extraction

Schedule C<bin/extract-attachments> to run at regular intervals via
cron.  For instance, the following F</etc/cron.d/rt> entry will run it
daily, which may be good to concentrate network or disk usage to times
when RT is less in use:

    0 0 * * * root /opt/rt4/local/plugins/RT-Extension-ExternalStorage/bin/extract-attachments

=back

=head1 CONFIGURATION

This module comes with a number of possible backends; see the
documentation in each for necessary configuration details:

=over

=item L<RT::Extension::ExternalStorage::Disk>

=item L<RT::Extension::ExternalStorage::Dropbox>

=item L<RT::Extension::ExternalStorage::AmazonS3>

=back

=head1 CAVEATS

This extension is not currently compatibile with RT's C<shredder> tool;
attachments which are shredded will not be removed from external
storage.

=cut

our $BACKEND;
our $WRITE;
$RT::Config::META{ExternalStorage} = {
    Type => 'HASH',
    PostLoadCheck => sub {
        my $self = shift;
        my %hash = $self->Get('ExternalStorage');
        return unless keys %hash;
        $hash{Write} = $WRITE;
        $BACKEND = RT::Extension::ExternalStorage::Backend->new( %hash );
    },
};

sub Store {
    my $class = shift;
    my $content = shift;

    my $sha = Digest::SHA::sha256_hex( $content );
    my ($key, $msg) = $BACKEND->Store( $sha => $content );
    return ($key, $msg) unless defined $key;
    return ($key);
}


package RT::Record;

no warnings 'redefine';
my $__DecodeLOB = __PACKAGE__->can('_DecodeLOB');
*_DecodeLOB = sub {
    my $self            = shift;
    my $ContentType     = shift || '';
    my $ContentEncoding = shift || 'none';
    my $Content         = shift;
    my $Filename        = @_;

    return $__DecodeLOB->($self, $ContentType, $ContentEncoding, $Content, $Filename)
        unless $ContentEncoding eq "external";

    unless ($BACKEND) {
        RT->Logger->error( "Failed to load $Content; external storage not configured" );
        return ("");
    };

    my ($ok, $msg) = $BACKEND->Get( $Content );
    unless (defined $ok) {
        RT->Logger->error( "Failed to load $Content from external storage: $msg" );
        return ("");
    }

    return $__DecodeLOB->($self, $ContentType, 'none', $ok, $Filename);
};

package RT::ObjectCustomFieldValue;

sub StoreExternally {
    my $self = shift;
    my $type = $self->CustomFieldObj->Type;
    my $length = length($self->LargeContent || '');

    return 0 if $length == 0;

    return 1 if $type eq "Binary";

    return 1 if $type eq "Image" and $length > 10 * 1024 * 1024;

    return 0;
}

package RT::Attachment;

sub StoreExternally {
    my $self = shift;
    my $type = $self->ContentType;
    my $length = $self->ContentLength;

    return 0 if $length == 0;

    if ($type =~ m{^multipart/}) {
        return 0;
    } elsif ($type =~ m{^(text|message)/}) {
        # If textual, we only store externally if it's _large_ (> 10M)
        return 1 if $length > 10 * 1024 * 1024;
        return 0;
    } elsif ($type =~ m{^image/}) {
        # Ditto images, which may be displayed inline
        return 1 if $length > 10 * 1024 * 1024;
        return 0;
    } else {
        return 1;
    }
}

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

All bugs should be reported via email to

    L<bug-RT-Extension-ExternalStorage@rt.cpan.org|mailto:bug-RT-Extension-ExternalStorage@rt.cpan.org>

or via the web at

    L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-ExternalStorage>.

=head1 COPYRIGHT

This extension is Copyright (C) 2009-2015 Best Practical Solutions, LLC.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

1;
